### PR TITLE
send the bill through email

### PR DESCRIPTION
**Pull request by mAbadsa**
Relates #157

### CODE REVIEW

**Review by Ahmed-aed — DISMISSED** [review]

File: server/controller/orders/updateOrderState.js [hunk 1]
```diff
@@ -59,9 +64,10 @@ const updateOrderState = async (req, res, next) => {
         if (resourcesPrice < 0 || Number.isNaN(resourcesPrice)) {
           throw boomify(400, 'please enter resources price');
         }
-
-        providerPriceHour = (await getPriceHourProvider(providerId)).rows[0]
-          .price_hour;
+        const { rows: [providerData]} = await getProviderDataById(providerId);
+        // providerPriceHour = (await getPriceHourProvider(providerId)).rows[0]
+        //   .price_hour;
+        providerPriceHour = providerData.price_hour;
```
Comment (mariamisa) [inline server/controller/orders/updateOrderState.js]: * you cant declare variables inside the switch so you need to declare ``` let providerData ``` outside the switch then give it the provider data.
```js
   [providerData] = (await getProviderDataById(providerId)).rows;

```
*  remove the comment code
* delete unused query file(getPriceHourProvider)  inside the require .

File: server/controller/orders/updateOrderState.js [hunk 2]
```diff
@@ -82,6 +88,19 @@ const updateOrderState = async (req, res, next) => {
...
+          `${userData}, ${providerData.email}`,
+          'Order Bill',
+          {
+            total: 150,
+            hourPrice: 15,
+            hourNumber: 10,
+            description: 'asd mjdj jhfdj jdhjhf jhjhf',
+            client: 'Muhammad',
+            provider: 'Ali',
+          },
```
Comment (mariamisa) [inline server/controller/orders/updateOrderState.js]: * I think the content of the email should be dynamic data according to the order and users information
* add the resource price to the msg content
```js
{
            total: newData.bill,
            hourPrice: providerData.price_hour,
            hourNumber: order.hour_number,
            description: order.description,
            client: userData.username,
            providerTitle: providerData.title,
            resoursePrice : newData.bill,
          },
```

File: server/utils/sendBill.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
+const nodemailer = require('nodemailer');
+
+const sendTheBill = async (
+  to,
+  subject,
+  { total, hourPrice, hourNumber, description, client, provider },
```
Comment (mariamisa) [inline server/utils/sendBill.js]: more clear to pass the object of email content and destructure the value , like this :
```js
const sendTheBill = async (to, subject, content, next) => {
  try {
    const {
      total,
      hourPrice,
      hourNumber,
      description,
      client,
      providerTitle,
    } = content;
```

File: server/utils/sendBill.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  next
+) => {
+  try {
+    const smtpTransport = nodemailer.createTransport({
+      service: 'Gmail',
+      auth: {
+        user: process.env.NODEMAILER_SENDER_EMAIL,
+        pass: process.env.SENDER_EMAIL_PASSWORD,
```
Comment (mariamisa) [inline server/utils/sendBill.js]: I like that you hide your email information, for more improvement destruct the values from process.env
```js
    const { NODEMAILER_SENDER_EMAIL, SENDER_EMAIL_PASSWORD } = process.env;

```

File: server/utils/sendBill.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+      <div style="padding: 10px; background-color: #25D595; color: #fff"><span>INVOICE TO: </span>${client}</div>
+      <div style="padding: 10px; background-color: #25D595; color: #fff"><span>Date: </span>${new Date().toDateString()}</div>
+      <hr/>
+      <table style="border:none;border-collapse:collapse;border-color:#aaa;border-spacing:0;">
+      <thead>
+        <tr>
+          <th style="text-align:left;vertical-align:top; background-color:#f38630;border-color:#aaa;border-style:solid;border-width:0px;color:#fff;
+          font-family:Arial, sans-serif;font-size:14px;font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;"><span style="font-weight:bold;">Client</span></th>
+          <th style="text-align:left;vertical-align:top; background-color:#f38630;border-color:#aaa;border-style:solid;border-width:0px;color:#fff;
+          font-family:Arial, sans-serif;font-size:14px;font-weight:normal;overflow:hidden;padding:10px 5px;word-break:normal;"><span style="font-weight:bold">Provider</span></th>
```
Comment (mariamisa) [inline server/utils/sendBill.js]: refactor the html content, remove all the duplicate style
* if you need to use colors inside the email, please use our site colors

File: server/controller/orders/updateOrderState.js [hunk 6]
```diff
@@ -21,6 +23,9 @@ const updateOrderState = async (req, res, next) => {
     });
 
     const [order] = rows;
+    const userData = await getUserData({
+      userId: order.user_id,
+    });
 
     if (!order) {
```
Comment (mariamisa) [inline server/controller/orders/updateOrderState.js:26]: refactor the condition to 
```js
if (!order) {
      throw boomify(404, 'There is no order');
    } else if (order.state === state || order.state === 'Finished') {
      throw boomify(400, `the order is already ${order.state}`);
    } else if (order.state === 'Accepted' && state !== 'Start') {
      throw boomify(409, 'the order is not started yet');
    }
```

**Review by mariamisa — CHANGES_REQUESTED** [review]
change the msg on the finish state to :
```js
res.json({
      statusCode: 200,
      message:
        state === 'Finished'
          ? 'your order completed successfully,check you email you will receive an email contains your order bill'
          : ` orders ${state}`,
    });
```

File: server/controller/orders/updateOrderState.js [hunk 7]
```diff
@@ -82,6 +88,19 @@ const updateOrderState = async (req, res, next) => {
         });
 
         await updateOrderRequestState(order.orders_request_id);
+        sendTheBill(
+          `${userData}, ${providerData.email}`,
```
Comment (mariamisa) [inline server/controller/orders/updateOrderState.js]: test your email and check if you pass the right data for first email 
```js
          `${userData.email}, ${providerData.email}`,

```

File: server/controller/orders/updateOrderState.js [hunk 8]
```diff
@@ -21,6 +23,9 @@ const updateOrderState = async (req, res, next) => {
     });
 
     const [order] = rows;
+    const userData = await getUserData({
+      userId: order.user_id,
+    });
```
Comment (mariamisa) [inline server/controller/orders/updateOrderState.js]: * hete you store the all result from query , destruct the rows to access user data 
```js
const { rows: result } = await getUserData({
      userId: order.user_id,
    });
    const [userData] = result;
```

**Review by mariamisa — CHANGES_REQUESTED** [review]

**Pull request merged**